module/trie.c

```diff
@@ -0,0 +1,271 @@
+    struct trie *parent;
+        // parent trie for traversing backwards
+};
+
+/*
+    Creates and allocates memory for new trie.
+    
+    Parameters:
+     - current: A char for the current character
+    
+    Returns:
+     - A pointer to the trie, or REDISMODULE_ERR if a pointer 
+       cannot be allocated
+*/
+struct trie *trie_new (char current)
+{
+    struct trie *t = RedisModule_Calloc(1,sizeof(struct trie));
+    if (t == NULL){
+        fprintf(stderr, "Could not allocate memory for trie\n");
+        return NULL;
+    } 
+    t->current = current;
+    t->children = RedisModule_Calloc(256,sizeof(struct trie *));
+    if (t->children == NULL){
+        fprintf(stderr, "Could not allocate memory for t->children\n");
+        return NULL;
+    }
+    t->is_word = 0;
+    t->parent = NULL;
+    return t;
+}
+
+/*
+    Free an entire trie.
+    Parameters:
+     - t: A trie pointer
+    
+    Returns:
+     - Always returns 0
+*/
+int trie_free(struct trie *t)
+{
+    if (t != NULL) {
+        for (int i=0; i<256; i++ ){
+            if (t->children[i] !=NULL)
+                trie_free(t->children[i]); // Called recursively because the entire trie and all of a trie's children are RedisModule_Calloc'ed
+        }
+        RedisModule_Free(t); // Used because the data structures are originally RedisModule_Calloc'ed
+    }
+    return 0;
+}
+
+/*
+    Creates new node in trie.
+    Parameters:
+     - current: A char indicating the character of the node being added
+     - t: A pointer to the trie where the node is to be added
+    
+    Returns:
+     - 0 on success, 1 if an error occurs.
+    Details: 
+     - Set t->children[current] to be current
+     - is_word for new node set to 0.
+*/
+int add_node(struct trie *t, char current)
+{
+    assert(t != NULL);
+    unsigned c = (unsigned) current; // Current is casted because the compiler will throw unnecessary warnings otherwise
+    if (t->children[c] == NULL)
+        t->children[c] = trie_new(current);
+    return 0;  
+}
+
+/*
+    Inserts word into trie.
+    Parameters:
+     - word: An char array to be inserted into the given trie
+     - t: A pointer to the given trie
+    
+    Returns:
+     - 0 on success, 1 if error occurs.
+    
+    Details:
+     - For each trie, check if entry of the next character exists in the children array:
+          - If so, move into that node in the array
+          - If not, add a new node and move into that node in the array
+     - Then move on to the next character in string
+     - Set the is_word of the last node to 1
+*/
+int trie_insert_string(struct trie *t, char *word)
+{
+    assert(t != NULL);
+    if (*word == '\0'){
+        t->is_word=1;
+        return 0;
+    } else {
+        char curr = *word;
+        int rc = add_node(t, curr);
+        if (rc != 0){
+            fprintf(stderr, "Fail to add_node");
+            return 1;
+        }
+        word++;
+        return trie_insert_string(t->children[(unsigned)curr], word);
```

Comment:
As you did in `add_node`, let's take the cast out of the array index to make this line more readable.